sponge.go

```diff
@@ -45,57 +73,48 @@ func (h HackerNewsItem) getFormatted() Formatted {
 		Body: fmt.Sprintf("Title: %s\nUrl: %s", h.Title, h.Url)}
 }
 
-func getHackerNews() []Formatted {
+func getHackerNews() ([]Formatted, error) {
 	hackerNewsListUrl := "https://hacker-news.firebaseio.com/v0/topstories.json"
 	hackerNewsItemUrl := "https://hacker-news.firebaseio.com/v0/item/%d.json"
 
-	client := &http.Client{Timeout: 10 * time.Second}
-	resp, err := client.Get(hackerNewsListUrl)
-	if err != nil {
-		fmt.Println(err)
-		return make([]Formatted, 0)
-	}
-	defer resp.Body.Close()
-
-	hnl := make([]int, 0)
-	decoder := json.NewDecoder(resp.Body)
-	decoder.Decode(&hnl)
+	var hnl []int
+	getJsonResponse(hackerNewsListUrl, &hnl)
 
 	// take only top items (returns 500 initially)
-	hnTopItems := make([]int, itemsToFetch)
-	copy(hnTopItems, hnl[:itemsToFetch])
-
-	hnTopItemsDetails := make([]Formatted, 0)
+	var hnTopItems []int
+	hnTopItems = append(hnTopItems, hnl[:*itemsToFetch]...)
+	var hnTopItemsDetails []Formatted
+	collectorChan := make(chan Formatted)
 
 	var wg sync.WaitGroup
-	wg.Add(itemsToFetch)
+	wg.Add(len(hnTopItems))
+
+	go func() {
+		wg.Wait()
+		close(collectorChan)
+	}()
 
 	for _, id := range hnTopItems {
-		go func(id int) {
+		id := id // need this or will only use LAST value of id for all goroutines
+		go func() {
 			defer wg.Done()
 
-			resp, err := client.Get(fmt.Sprintf(hackerNewsItemUrl, id))
+			hnItemUrl := fmt.Sprintf(hackerNewsItemUrl, id)
+			item := HackerNewsItem{}
+			err := getJsonResponse(hnItemUrl, &item)
 			if err != nil {
 				fmt.Println("error!", err)
```

Comment:
instead of this, you can also fmt.Printf("%#v\n", err) and you'll get the type of the error and anything else interesting about it.  "error!" isn't really super useful ;)  with %#v, it will likely even print the word error!